src/main/java/hudson/matrix/MatrixProject.java

```diff
@@ -199,14 +201,35 @@
      */
     private String childCustomWorkspace;
 
+    /**
+     * Lock to prevent project changes on different build at the same time
+     */
+    private ReadWriteLock buildLock;
```

Comment:
`buildLock = new ReentrantReadWriteLock();` should be enough as transient fields will be null.